inference-engine/thirdparty/clDNN/src/graph_optimizer/remove_redundant_reorders.cpp

```diff
@@ -155,9 +160,24 @@ void remove_redundant_reorders::run(program_impl& p) {
 
         auto& r_node = node->as<reorder>();
 
-        bool no_output_optimization = remove_output_reorders ?
-            r_node.is_output() && (r_node.get_dependency(0).is_output() || r_node.get_dependency(0).is_type<input_layout>() ||
-                r_node.get_dependency(0).can_be_optimized() || r_node.get_dependency(0).get_users().size() != 1) : r_node.is_output();
+        auto& dep0 = r_node.get_dependency(0);
+        // We shouldn't optimize output when node before reorder is argmax or output from argmax.
+        bool direct_dep_check = dep0.is_type<arg_max_min>() ||
+            dep0.is_type<ctc_greedy_decoder>() ||
+            dep0.is_type<gather_nd>() ||
+            dep0.is_type<reduce>();
+        bool second_order_dep_check = dep0.get_dependencies().size() > 0;
+        if (second_order_dep_check) {
+            auto& second_dep0 = dep0.get_dependency(0);
+            second_order_dep_check &= (second_dep0.is_type<arg_max_min>() ||
+                                       second_dep0.is_type<ctc_greedy_decoder>());
```

Comment:
Why? These two conditions abode are far from being obvious. To me it seems to be a workaround rather than good condition. Why `direct_dep_check` checks these 4 layer types? Why `second_order_dep_check` checks another set of types? Why we don't need {third, fourth, ...}_order_dep_check? 
So this part is pretty fragile to my mind.

Reply:
Need to rework this code to keep output reorder optimization...
Need to add output/primitive memory mapping to correctly handle setting output memory for such optimized out primitives...

Reply:
Rewrote all the code that works with optimized out layers and mutable_data in updated version